File: Day4.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Threading;

namespace adventofcode2017
{
    class Day4
    {
        internal static bool IsValid(string password)
        {
            var raw = password.Split(' ');
            return raw.Length == raw.ToHashSet().Count;                    
        }

        internal static void Run1()
        {
            var input = File.ReadAllLines("input/Day4.txt");
            var count = 0;
            foreach (var password in input)
            {
                if (IsValid(password))
                    count++;
            }

            Console.WriteLine(count);
        }        

        internal static bool IsValid2(string password)
        {
            var raw = password.Split(' ');
            return raw.Length == raw.Select(x => string.Concat(x.OrderBy(c => c)))
                                    .ToHashSet()
                                    .Count;                    
        }

        internal static void Run2()
        {
            var input = File.ReadAllLines("input/Day4.txt");
            var count = 0;
            foreach (var password in input)
            {
                if (IsValid2(password))
                    count++;
            }

            Console.WriteLine(count);
            Console.WriteLine(IsValid2("oiii ioii iioi iiio"));
        }        

    }

}

File: src/bin/day4.rs
const INPUT: &'static str = include_str!("../../input/Day4.txt");

fn main() {
    day1();
}

fn day1() {
    run1();
    run2();
}

fn run1() {
    let mut count = 0;
    for password in INPUT.lines() {
        if is_valid(password) {
            count += 1;
        }
    }
    println!("{}", count);
}

fn is_valid(password: &str) -> bool {
    let mut v: Vec<&str> = password.split(' ').collect();
    let len_orig = v.len();
    v.sort();
    v.dedup();
    len_orig == v.len()
}

fn run2() {
    let mut count = 0;
    for password in INPUT.lines() {
        if is_valid_2(password) {
            count += 1;
        }
    }
    println!("{}", count);
}

fn is_valid_2(password: &str) -> bool {
    let v: Vec<&str> = password.split(' ').collect();
    let len_orig = v.len();
    let mut v: Vec<String> = v.iter().map(|s| {
        let mut w = s.chars().collect::<Vec<char>>();
        w.sort();
        w.iter().collect()
    }).collect();
    v.sort();
    v.dedup();
    len_orig == v.len()
}
